### DISCUSSION TITLE
debug kube-vip locally

### DISCUSSION THREAD

**Discussion by laurijssen** [kube-vip/kube-vip#959] [Q&A]
I am interested in working on kube-vip so cloned the repo and to get a feel on the structure of the code I started debugging from vscode with launch.json like this:

```
    "configurations": [
        {
            "name": "Launch Package",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "cwd": "${workspaceFolder}",
            "program": "${workspaceFolder}/main.go",
            "args": [ "manager", "--arp" ],
            "env": {
                "cp_enable": "true",
                "vip_interface": "eth1",
                "address": "192.168.100.200"
            }
        }
    ]
```

This starts kube-vip like:

```
cp_enable="true" vip_interface="eth1" address="192.168.100.200" ./kube-vip manager --arp
```

but once the actual net interface is created it errors on permission denied on the following line:

```
func (configurator *network) AddIP() error {
    if err := netlink.AddrReplace(configurator.link, configurator.address); err != nil {
        return errors.Wrap(err, "could not add ip")
    }
    ....
}```

Of course, as I dont have permissions, it works with root. But how can we debug kube-vip as a normal user? Setting caps on what?

Or how is this usually done?

**Comment by Cellebyte** [comment]
@laurijssen first of all, thanks that you are looking into the Codebase.

To manipulate network you need at least the CAP_NET_ADMIN capability and for ARP to work you additionally need the capability CAP_NET_RAW. You could first create a network namespace locally to not manipulate network configuration of your host system. This would require root for creation of this network namespace. Then you can change context into this network namespace and assign your binary the capabilities needed.


`sudo setcap 'cap_net_admin=ep cap_net_raw=ep' bin/manager`

To simplify this setup you can build a container image and run it with the capabilities for testing.

Otherwise you can create a local kubernetes cluster using kind. If you want to test kubernetes functionality.

If you have more questions just ask.

**Reply by laurijssen** [reply]
Thanks! Been testing with network namespaces but got stuck at the "switch context to the network namespace"

```
sudo ip netns add kubevip
sudo ip link set veth0 netns kubevip
sudo ip netns exec kubevip bash
```

This switches context but then still the debugger emits operation not supported.

![afbeelding](https://github.com/user-attachments/assets/b1cb4aed-197a-407d-b813-89e1722cec4a)

This is for stepping through the code locally, it does work from a k8s cluster or container

**Reply by Cellebyte** [reply]
you need to call the veth interface inside the netns `eth1` or did you change the config accordingly.
